New behaviour based on sub-themes

marionb opened the pull request:
closes #1196 
TODO:
- [x] implement new sub-themes on the level of themes
- [x] add test data
- [x] fix json & xml extracts
- [x] Fix Print
- [x] implement sorting for themes and sub-themes
- [x] check if this sorting is what is expected
- [x] fix current tests
- [ ] add tests if needed

@vvmruder and @voisardf Currently I am sorting the themes and sub-themes according to their `extract_index` and no longer with what was implemented earlier. Before I clean out the old sort mechanism can you tell me if the is fine like this or if we need something like the old sorting mechanism again?
Also the sorting is now done in the extract together with the sorting of the `law status`, so that we have to sort this part only once.

voisardf commented:
Yes, sorting by the given index is the new way it should work.

Sorting principles are described on page 8 ff of the static extract specs: https://github.com/openoereb/pyramid_oereb/files/6442562/OEREB-Data-Extract_de.pdf and detailed on page 13, first line.
Interestingly enough, the sortindex sems not to appear in the extract model and from the specifications it is not exactly clear if it should be passed as integer or string: https://github.com/openoereb/models/issues/3

marionb commented:
thanks @voisardf I removed the old sorter logic.
Currently I implemented the index thinking it was an integer (I think it will also work wit a string). I will leave it up to the person who integrates the data to set the index correctly so that the sorting is in accordance with the spec of the the static extract.

vuilleumierc reviewed (commented):
Tested locally and it works well, thanks for putting all this together!
Great that you added new geometries (points and lines with the new LandUsePlans PLRs), although I am surprised that in the static extract we only ever see the same rectangle, although "Anteil" has sometimes different units (m^2, m, -). Do you know why that is? I don't think it is related to your PR though as it as always been the case.
Otherwise I just have a few comment in the code

voisardf commented:
I also tested the branch locally and got the extract (json/xml/pdf). The pdf only when commenting the line 8 regarding pdf conformance PDF/1-a. Which is still an open issue regarding the logos. So that seems fine to me.
I also get seven failing tests. Most of them schema-validation stuff. Do you have the same? @vuilleumierc @marionb

vuilleumierc commented:
Yes I also get the usual seven failing tests:
```
FAILED tests/contrib/print_proxy/test_mapfish_print.py::test_mapfish_print_entire_extract
FAILED tests/renderer/test_xml.py::test_extract_against_schema[parameter0-test_extract0] - lxml.etree.XMLSchemaParseError: Element '{http://www.w3.org/2001/XMLSchema}import':...
FAILED tests/renderer/test_xml.py::test_extract_against_schema[parameter1-test_extract1] - lxml.etree.XMLSchemaParseError: Element '{http://www.w3.org/2001/XMLSchema}import':...
FAILED tests/renderer/test_xml.py::test_extract_against_schema[parameter2-test_extract2] - lxml.etree.XMLSchemaParseError: Element '{http://www.w3.org/2001/XMLSchema}import':...
FAILED tests/xtf_import/test_legend_entry.py::test_parse - TypeError: 'topic' is an invalid keyword argument for LegendEntry
FAILED tests/xtf_import/test_public_law_restriction.py::test_parse - TypeError: 'legend_text' is an invalid keyword argument for PublicLawRestriction
FAILED tests/xtf_import/test_federal_topic.py::test_load - SystemExit: 1
```

marionb commented:
@voisardf Yes I also have these test that fail. As far as I saw they also fail on master and have dedicated issues so I do not see is as: in the scope of this PR, to fix them -> https://github.com/openoereb/pyramid_oereb/wiki/DEV-Time-contracts

The PDF that fails it is also correct, unless you remove the 3 lines.

vvmruder commented:
If it was approved an tested by @voisardf you can merge it. I don't have the time today to test it :)

vuilleumierc reviewed (approved):
Looks very good, thanks!

marionb commented:
Any one has a clue how to fix this codacity issue? I am stating to think that codacity is actually wrong - or the error is a completely different one than what is stated.

voisardf reviewed (commented):
Code looks good to me. JSON/XML data extract work, but I could not generate a PDF from this branch. Normal? Or do I have to be  on a particular mpf-branch?

marionb commented:
@voisardf thanks for testing and reviewing. To generate the PDF you can use the branch https://github.com/openoereb/pyramid_oereb_mfp/tree/Turn_Off_PDFA
The current master does not work because some commits are missing in this PR. Once merging this in to master it will work again with the OEREB_MFP master.

marionb commented on pyramid_oereb/standard/templates/plr_integer_primary_keys.py.mako:
```diff
@@ -211,7 +211,7 @@ class PublicLawRestriction(Base):
         legend_text (dict): The multilingual textual representation of the public law restriction.
         topic (str): Category for this public law restriction (name of the topic).
-        sub_theme (dict): Multilingual textual explanation to subtype the topic attribute.
+        sub_theme (str): Multilingual textual explanation to subtype the topic attribute.
```
I removed the file :-)

marionb commented on docker/config.yaml.tmpl:
```diff
@@ -82,9 +82,6 @@ vars:
       # Whether to display the Certification section in the pdf extract or not.
       # Default to true
       display_certification: true
```
Good idea if there is only one canton using this.
I changed it.

voisardf reviewed (approved):
I was not able to get the PDF yet, I still get _Ein oder mehrere ÖREB-Themen stehen momentan nicht zur Verfügung. Daher kann kein Auszug erstellt werden. Versuchen Sie es zu einem späteren Zeitpunkt erneut. Wir entschuldigen uns für die Unannehmlichkeiten._

but if you need the PR to advance, ok for me to merge it while I try to get a better idea where the mfp error lies

marionb commented:
@voisardf I'll re-base the PR on to master. It will make testing easier.
